rust/src/passwd.rs

```diff
@@ -567,3 +634,299 @@ impl PasswdDB {
+                let buf_rd = BufReader::new(old_passwd_stream);
+                let entries = nameservice::passwd::parse_passwd_content(buf_rd)?;
+                for user in entries {
+                    self.users.insert(
+                        user.name,
+                        (Uid::from_raw(user.uid), Gid::from_raw(user.gid)),
+                    );
+                }
+            }
+            CheckPasswd::Data(data) => {
+                for user in &data.entries {
+                    self.users.insert(user.0.clone(), user.1.ids());
+                }
+            }
+        };
+
+        Ok(())
+    }
+
+    #[context("Rendering group entries from treefile check-groups")]
+    fn populate_groups_from_treefile(
+        &mut self,
+        treefile: &mut Treefile,
+        repo_previous_rev: &Option<(&ostree::Repo, &str)>,
+    ) -> Result<()> {
+        let config = treefile.parsed.get_check_groups();
+
+        match config {
+            CheckGroups::None => {}
+            CheckGroups::File(f) => {
+                let fp = treefile.externals.group_file_mut(f)?;
+                let buf_rd = BufReader::new(fp);
+                let entries = nameservice::group::parse_group_content(buf_rd)?;
+                for group in entries {
+                    self.groups.insert(group.name, Gid::from_raw(group.gid));
+                }
+            }
+            CheckGroups::Previous => {
+                // This logic short-circuits if there is no previous commit, or if
+                // it doesn't contain a group file. Nothing to validate in that case.
+                let (repo, previous_rev) = match repo_previous_rev {
+                    Some(v) => v,
+                    None => return Ok(()),
+                };
+                let (prev_root, _name) = repo.read_commit(previous_rev, gio::NONE_CANCELLABLE)?;
+                let old_path = match prev_root.resolve_relative_path("usr/lib/group") {
+                    Some(v) => v,
+                    None => return Ok(()),
+                };
+                let old_passwd_stream = old_path.read(gio::NONE_CANCELLABLE)?.into_read();
+                let buf_rd = BufReader::new(old_passwd_stream);
+                let entries = nameservice::group::parse_group_content(buf_rd)?;
+                for group in entries {
+                    self.groups.insert(group.name, Gid::from_raw(group.gid));
+                }
+            }
+            CheckGroups::Data(data) => {
+                for (groupname, gid) in &data.entries {
+                    let id = Gid::from_raw(*gid);
+                    self.groups.insert(groupname.clone(), id);
+                }
+            }
+        };
+
+        Ok(())
+    }
+
+    #[context("Validating user entries according to treefile check-passwd")]
+    fn validate_treefile_check_passwd(
+        &self,
+        old_subset: &PasswdEntries,
+        rootfs: i32,
+        ignored_users: &[String],
```

Comment:
Since we're calling `.contains()` on this in a loop, shouldn't the caller pass it as a `BTreeSet<&String>`?  This would also allow using `.unwrap_or_default()` instead of passing a `&vec[]!`.  Or maybe it's an `Option<BTreeSet<&String>>`.

Reply:
Now that you highlight it, I see it's indeed quite sloppy. I'll try to improve it that way.